### PR TITLE
feat(color): closed-form two-white (RGBWW) allocation (P7, #4198)

### LINKED ISSUE #4198: P7: two-white (RGBWW) allocation is unimplemented — findings from a failed attempt
P7 (#4041) asks for the C3 allocation policy on "RGBW/RGBWW". The one-white case is done and merged (#4188, #4193). Two whites are not, and this records what I established attempting it, including where I went wrong, so the next attempt starts further along.

## Why two whites is a different problem

With one white the feasible levels form an interval and the policy picks an end. With two, maximizing `w1 + w2` is a linear program over a polygon.

`docs/color-gamut-algorithm-selection.md` already records that the obvious reduction — run the one-white form per white, keep the better answer — agrees with the reference on **every** corpus vector while being wrong on **85%** of random reachable targets, because a single emitter's drive is capped at 1.

## A reduction that works, up to a point

At a **fixed total** `s = w1 + w2`, substituting `w2 = s − w1` gives RGB drives of

```
(d0 − s·dW2) − w1·(dW1 − dW2)
```

which is exactly the one-white shape with a shifted target and a difference column. So testing a total costs the same six comparisons as testing a single white level, and the polygon never has to be enumerated.

Checked against an exact vertex enumeration: **worst disagreement 1.8e-12** over the targets where both methods returned an answer.

At the chosen total the split is still free, and the P5 reference settles it by minimizing the sum of squares of the RGB drives. Those are affine in `w1`, so that is a quadratic — one division, not a search. With the split included, the method reproduces the reference on **all 48** `rgbww` corpus vectors to 1.7e-12.

## Where it breaks, and the mistake I made

The achievable totals form an **interval** `[s_lo, s_hi]`, not `[0, s_hi]`. A bright target is infeasible at `s = 0` — the primaries alone cannot reach it — so a bisection seeded at zero has no feasible starting point and reports the target unreachable.

My validation hid this. The harness skipped any target where my method returned no answer, and I read "0 disagreements over 10,285 targets" as success without noticing that **roughly 30,000 of 40,000 were skipped**, many of them precisely the cases the method could not handle. The corpus did not catch it either: its `rgbww` vectors are all reachable with the primaries alone or nearly so, which is the same corpus gap that hides the 85% figure above.

## What a working version needs

The feasible-`s` interval has to be found, not assumed to start at zero. All the bounds are **affine in `s`**, so `max(lower bounds) <= min(upper bounds)` is a conjunction of pairwise linear inequalities in `s` — roughly 25 of them for 5 lower and 5 upper bounds — each solvable for an `s` bound. Intersecting them gives `[s_lo, s_hi]` in closed form, with no search at all. That looks more promising than either the vertex enumeration or a two-phase scan-then-bisect, but I have not implemented or measured it.

Cost is the open question either way: ~25 divisions per pixel against six for the one-white case, and the one-white path is already the only per-pixel division in the pipeline (see the note in `white_allocation.h`).

## Also needed

Corpus vectors bright enough to require both whites. Without them, nothing about this can be validated against the P5 reference — only against an independently written solver, which is weaker evidence.

Related: #4041, #4188, #4193, #4197.

### PR DESCRIPTION
**Pull request by zackees**
Closes #4198. Part of #4041 (P7), under #4034.

P7 asks for the C3 allocation policy on "RGBW/RGBWW". One white landed in #4188
and #4193. Two did not, and #4198 recorded a failed attempt at them.

## The closed form

Write the total `s = w₁ + w₂` and substitute `w₂ = s − w₁`. The RGB drives
become

```
(d₀ − s·dW₂) − w₁·(dW₁ − dW₂)
```

which is the one-white shape with a shifted target and a difference column. So
at any fixed total the feasible `w₁` is again an interval, bounded by five
lower and five upper bounds — three from the RGB drives, and two more because
`w₁` and `w₂ = s − w₁` are drives in their own right.

Every one of those bounds is **affine in `s`**. So "some split exists at this
total" is exactly the conjunction of the pairwise inequalities
`lowerₖ(s) ≤ upperⱼ(s)`, each linear in `s` and each solvable for one `s`
bound. Intersecting the 25 of them gives the feasible totals directly — no
bisection, no vertex enumeration, nothing iterative on the per-pixel path
(A3/B11).

**The interval has to be found, not assumed to start at zero.** That is where
the previous attempt broke: a bright target is unreachable with the primaries
alone, so its feasible totals begin above zero and a bisection seeded at zero
reports it out of gamut.

## Measured, with nothing skipped

The previous attempt reported "0 disagreements over 10,285 targets" while its
harness had silently dropped roughly 30,000 of 40,000 — including precisely
the bright targets it could not handle. So every sample here is classified by
both methods, and a disagreement about *feasibility* counts as loudly as a
disagreement about the answer:

| targets | how drawn | solved by both | disagreements |
|---|---|---:|---:|
| 8,000 | uniform drives | 8,000 | 0 |
| 4,000 | both whites near full | 4,000 | 0 |
| 4,000 | drives to 1.35×, many out of hull | 2,862 | 0 |

Worst difference in total white: 2.2 × 10⁻¹⁵. The corpus's 48 `rgbww` vectors
also reproduce to < 10⁻⁹.

One target in 4,000 sits on the hull to within 10⁻⁷ and is admitted by the
enumeration's per-constraint slack while the closed form's exact interval
refuses it. That is counted as its own category and asserted to stay rare —
not folded into agreement, and not skipped.

## Two things measurement decided

**The slack is one-sided.** The exact hull corner was refused: five emitters at
full drive solve, through a quantized matrix, to a target needing 57 raw units
more than a total of 2.0 can supply. The bounds therefore carry slack on the
`drive ≤ 1` side only — the `drive ≥ 0` side would be *spent* by a policy that
maximizes white, which is the same asymmetry `allocateEmitterDrivesQ16` already
argues for.

**And it is half the final check's.** Relaxing a bound and then re-checking the
same drive against the same number puts the boundary case exactly on the
threshold, where two independent roundings decide it — measured rejecting a
drive at 65601 against a limit of 65600, one raw unit, on the RGB-preferred end
of a target needing both whites.

## What was written and then deleted

The least-squares split rule #4198's plan called for. It is a quadratic in the
split, so one division, and it was implemented here. Measurement removed it:
**at the extreme total the feasible split is a single point**, so there is
nothing for any rule to choose. Width measured 0.0 over 4,000 targets on the
corpus's cool/warm device, and over 951 on a device built specifically so one
drive's constraint is parallel to `w₁ + w₂` — the shape that could have
produced an optimal edge.

Two whites of the *same* colour are the one case where the split really is
free, and there the reference's own tie-break — lexicographically smallest
drives — is the end this takes anyway.

Shipping an untestable division on a per-pixel embedded path to serve an
unreachable case is the wrong trade, so it is gone and the reason is recorded
in the header and in `docs/color-gamut-algorithm-selection.md`.

## Tests

`ci/tests/test_color_rgbw_study.py` gates the algorithm against the reference's
enumeration (the sweeps above). `tests/fl/gfx/white_allocation.cpp` gates the
s16.16 implementation against an independent float vertex enumeration written
in the test — reproduction, maximal total, targets one white cannot reach,
out-of-hull rejection, the RGB-preferred end, the degenerate equal-whites
device, and the claim that the split interval is a point.

Eight mutations of the implementation were run; all eight fail the tests:

| mutation | caught by |
|---|---|
| always take the lowest total | maximal-total, both-whites |
| drop the total's lower bound (fixed-drive branch) | equal-whites RGB-preferred |
| pairwise bound always narrows the top | maximal-total |
| ignore the sign of the difference column | maximal-total |
| `difference = per_white1` | reproduction, maximal-total |
| take the high end of the split | equal-whites split |
| drop `w₁ ≥ s − 1` | maximal-total |
| drop `w₁ ≤ s` | the Python gate (3 sweeps) |

`bash test --cpp --clean`: 383 passed, 0 failed (299 unit, 84 examples).
`bash lint` clean.

## Still open

The golden corpus has no `rgbww` vector that needs both whites — all 48 are
reachable with one white or none — so it cannot distinguish a correct two-white
allocation from the reduction it replaces. The sweeps stand in for that,
measured against the reference's own enumeration rather than its recorded
output. Corpus vectors bright enough to need both whites would close the gap;
that is noted in the document rather than claimed as done.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

* **New Features**
  * Added support for allocating color targets across two white emitters.
  * Supports white-preferred and RGB-preferred allocation policies.
  * Provides consistent allocation for targets requiring both white emitters, including identical emitters.
  * Added validation for unreachable targets, invalid inputs, overflow, and degenerate configurations.

* **Documentation**
  * Documented the closed-form two-white allocation approach and implementation behavior.

* **Tests**
  * Added comprehensive coverage for target reconstruction, boundary conditions, rejection cases, identical emitters, and quantized output.

### CODE REVIEW

File: ci/color_rgbw_study.py [hunk 1]
```diff
@@ -203,3 +212,123 @@ def best_single_white(
...
+            # Expressed as a pair of coincident w1 bounds, so the pairwise
+            # sweep below picks the same totals up without a special case.
+            at_dark = (constant / slope, 0.0)
+            at_full = ((constant - 1.0) / slope, 0.0)
+            if slope > 0:
+                uppers.append((at_dark[0], 0.0))
+                lowers.append((at_full[0], 0.0))
+            else:
+                uppers.append((at_full[0], 0.0))
+                lowers.append((at_dark[0], 0.0))
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/color_rgbw_study.py]: Confirmed, and the measurement is worse than the description. Fixed in a3d3293.

You are right that a constant `w1` bound cannot express a bound on the total -- the value being computed there is a total, not a split, so the encoding was dimensionally wrong. On a device whose two whites are the same colour, which is the only shape that reaches the branch:

- before: 2206 of 3000 random reachable targets disagreed with `most_white_two` about feasibility, and where both answered the total was out by as much as 1.87;
- after: 3000 of 3000 solved, zero mismatches, worst total delta 2.2e-16 -- and the same on 1.35x targets that fall outside the hull.

`_split_bounds` now returns `low_total`/`high_total` alongside the `w1` bounds and narrows them directly in that branch, which is what the C++ already did: `allocateTwoWhiteDrivesQ16` narrows them in place, and a mutation removing that fails the C++ tests. So this was the Python model diverging from the implementation it exists to validate, not a bug in the shipped path.

The regression test you asked for is `test_it_agrees_when_the_two_whites_are_the_same_colour` -- a 3000-target sweep through both `allocate_two_white` and `most_white_two`. Restoring the old branch fails it. Nothing else in the file reaches that code, because every other device here has a non-zero difference in all three channels, which is exactly why it went unnoticed.

File: docs/color-gamut-algorithm-selection.md [hunk 2]
```diff
@@ -363,6 +363,61 @@ polygon; `most_white_two` solves it by exact vertex enumeration, and the
...
 simplifies to it later.
 
+### Two whites also have a closed form
+
+Not the reduction above, and not a search. Write the total `s = w₁ + w₂` and
+substitute `w₂ = s − w₁`: the RGB drives become
+
+```
+(d₀ − s·dW₂) − w₁·(dW₁ − dW₂)
+```
```
[the remark this answers is not in the document]
Comment (zackees) [reply docs/color-gamut-algorithm-selection.md]: Fixed in a3d3293 -- converted to the indented style the rest of the document uses, matching the `d(w)` formula above it.

File: ci/color_rgbw_study.py [hunk 3]
```diff
@@ -203,3 +212,151 @@ def best_single_white(
         if best is None or span.highest > best:
             best = span.highest
     return best
+
+
+@dataclass(frozen=True, slots=True)
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/color_rgbw_study.py:223]: Done, and applied to all four dataclasses in the file rather than only the new one. `WhiteLevelRange`, `WhitePreferredDrives` and `TwoWhiteLevels` were already undecorated, so decorating `_SplitBounds` alone would have left the file less consistent than it started.

I measured the cost before adding it, because `_SplitBounds` is built once per target and the sweeps build tens of thousands: 200,000 constructions take 0.090s undecorated and 0.092s decorated. There is nothing to trade off.
